File: bc-w3/bcw3/Car/Point.cpp
#include <iostream>
#include <cmath>
#include "Point.h"

Point::Point(double x, double y) : x(x), y(y) {}

Point::~Point() {}

double Point::getX() const {
    return x;
}

double Point::getY() const {
    return y;
}

double Point::distance(const Point& other) const {
    return hypot(other.x-this->x, this->y-other.y);
}

bool Point::operator==(const Point& other) const {
    return other.x == this->x && other.y == this->y; 
}

bool Point::operator!=(const Point& other) const {
    return !(*this == other);
}

std::ostream& operator<<(std::ostream& out, const Point& point) {
    return out << '(' << point.getX() << ", " << point.getY() << ')';
}

File: bc-w3/java/Car/Point.java
public class Point implements Cloneable {
    private double x;
    private double y;
    
    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }
    
    public Point() {
        this(0, 0);
    }
    
    public double getX() {
        return x;
    }
    
    public double getY() {
        return y;
    }
    
    public double distance(Point other) {
        return Math.hypot(other.x-this.x, this.y-other.y);
    }
    
    public boolean equals(Point other) {
        return other.x == this.x && other.y == this.y;
    }
    
    public String toString() {
        return "(" + this.x + ", " + this.y + ")";
    }
    
    public Point clone() throws CloneNotSupportedException {
        return (Point) super.clone();
    }
}
